client request can get lost when range is merged

The following test appears to have failed:

[#6707](https://circleci.com/gh/cockroachdb/cockroach/6707):

```
I0901 01:34:21.312058 1013 stopper.go:229  draining; tasks left:
1      storage/replica.go:1254
I0901 01:35:20.487935 1013 storage/replica.go:992  gossiping cluster id  from store 1, range 1
W0901 01:35:20.488024 1013 storage/store.go:593  error gossiping first range data: storage/replica.go:972: node is stopping
W0901 01:35:20.488566 1013 storage/store.go:612  error gossiping configs: storage/replica.go:972: node is stopping
panic: test timed out after 1m10s

goroutine 5387 [running]:
testing.startAlarm.func1()
	/usr/local/go/src/testing/testing.go:703 +0x132
created by time.goFunc
	/usr/local/go/src/time/sleep.go:129 +0x3a

goroutine 1 [chan receive, 1 minutes]:
testing.RunTests(0x1221c48, 0x174c580, 0xb9, 0xb9, 0x7de301)
	/usr/local/go/src/testing/testing.go:562 +0x8ad
--
	/go/src/github.com/cockroachdb/cockroach/storage/id_alloc.go:130 +0x93d
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc820428120, 0xc8204800d0)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:80 +0x52
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:81 +0x62
FAIL	github.com/cockroachdb/cockroach/storage	70.034s
=== RUN   TestBatchBasics
I0901 01:34:08.284690 981 storage/engine/rocksdb.go:122  closing in-memory rocksdb instance
--- PASS: TestBatchBasics (0.00s)
=== RUN   TestBatchGet
I0901 01:34:08.285482 981 storage/engine/rocksdb.go:122  closing in-memory rocksdb instance
--- PASS: TestBatchGet (0.00s)
=== RUN   TestBatchMerge
I0901 01:34:08.287797 981 storage/engine/rocksdb.go:122  closing in-memory rocksdb instance
--- PASS: TestBatchMerge (0.00s)
=== RUN   TestBatchProto

```
Please assign, take a look and update the issue accordingly.

--- a/storage/store.go
+++ b/storage/store.go
@@ -261,6 +261,7 @@ type Store struct {
 	scanner           *replicaScanner // Range scanner
 	feed              StoreEventFeed  // Event Feed
 	removeReplicaChan chan removeReplicaOp
+	proposeChan       chan proposeOp
 	multiraft         *multiraft.MultiRaft
 	started           int32
 	stopper           *stop.Stopper
@@ -368,6 +369,7 @@ func NewStore(ctx StoreContext, eng engine.Engine, nodeDesc *proto.NodeDescripto
 		uninitReplicas:    map[proto.RangeID]*Replica{},
 		nodeDesc:          nodeDesc,
 		removeReplicaChan: make(chan removeReplicaOp),
+		proposeChan:       make(chan proposeOp),
 	}
 
 	// Add range scanner and configure with queues.
@@ -1489,11 +1491,24 @@ func (s *Store) resolveWriteIntentError(ctx context.Context, wiErr *proto.WriteI
 	return wiErr
 }
 
+type proposeOp struct {
+	idKey cmdIDKey
+	cmd   proto.RaftCommand
+	ch    chan<- <-chan error
+}
+
 // ProposeRaftCommand submits a command to raft. The command is processed
 // asynchronously and an error or nil will be written to the returned
 // channel when it is committed or aborted (but note that committed does
 // mean that it has been applied to the range yet).
 func (s *Store) ProposeRaftCommand(idKey cmdIDKey, cmd proto.RaftCommand) <-chan error {
+	ch := make(chan (<-chan error))
+	s.proposeChan <- proposeOp{idKey, cmd, ch}
+	return <-ch
+}
+
+// proposeRaftCommandImpl runs on the processRaft goroutine.
+func (s *Store) proposeRaftCommandImpl(idKey cmdIDKey, cmd proto.RaftCommand) <-chan error {
 	value := cmd.Cmd.GetValue()
 	if value == nil {
 		panic("proposed a nil command")
@@ -1588,6 +1603,9 @@ func (s *Store) processRaft() {
 
 			case op := <-s.removeReplicaChan:
 				op.ch <- s.removeReplicaImpl(op.rep)
+
+			case op := <-s.proposeChan:
+				op.ch <- s.proposeRaftCommandImpl(op.idKey, op.cmd)
 
 			case <-s.stopper.ShouldStop():
 				return

